### ISSUE TITLE
Order of elements in datasource item XML is wrong

### ISSUE THREAD

**Issue by Bartman0** [tableau/server-client-python#335]
I wrote this issue report on the community forums, [(post):](https://community.tableau.com/message/800772#800772)

I am having trouble getting a data source published through the REST API.
The resulting error is 400011: Bad Request.
 
If I check the request against the 3.0 REST API XSD, it is valid:
 
> <tsRequest><datasource name="Test RKO"><connectionCredentials embed="true" name="HANSANDERS_WDR" password="********" /><project id="31ca5587-0b05-4abe-87fe-77ae31f153d7" /></datasource></tsRequest>
 
According to XML Spy at least.
 
I changed the request_factory.py to do this:

class DatasourceRequest(object):  
   def _generate_xml(self, datasource_item, connection_credentials=None, connections=None):  
  xml_request = ET.Element('tsRequest')  
  datasource_element = ET.SubElement(xml_request, 'datasource')  
  datasource_element.attrib['name'] = datasource_item.name  
  
   if connection_credentials is not None and connections is not None:  
   raise RuntimeError('You cannot set both `connections` and `connection_credentials`')  
  
   if connection_credentials:  
  credentials_element = ET.SubElement(datasource_element, 'connectionCredentials')  
  credentials_element.attrib['name'] = connection_credentials.name  
  credentials_element.attrib['password'] = connection_credentials.password  
  credentials_element.attrib['embed'] = 'true' if connection_credentials.embed else 'false'  
  
   if connection_credentials.oauth:  
  credentials_element.attrib['oAuth'] = 'true'  
  
   project_element = ET.SubElement(datasource_element, 'project')  
  project_element.attrib['id'] = datasource_item.project_id  

The fix is to have the project definition come after the credentials. Before this change, the request was not valid according to the XSD.
 
Overall, I am trying to achieve this:
- change database environment in the data source
- publish the data source with changed credentials
- I am using the document and service Python SDK's to do this

Script is attached.

[datasource_publish.py.zip](https://github.com/tableau/server-client-python/files/2341539/datasource_publish.py.zip)

**Comment by t8y8** [comment]
@irwando @RussTheAerialist the ordering of elements shouldn't matter -- this seems like a Server-side problem?

**Comment by Bartman0** [comment]
The ordering matters looking at the XSD.
But the server should not enforce that, that's what you are saying?


NB Vanaf mobiel, dus wees bewust van telegramstijl.


Richard Kooijman
Inergy
Pelmolenlaan 19  • 3447 GW Woerden • T: +31 348 457 666 • M: +31 631 673 387 • www.inergy.nl
4 oktober | Event l Inspiriment "Maak impact met data & analytics"
15 november | Event l Talking Dinner
Vraag | Waar is bij u waarde uit data te halen?
On 5 Sep 2018, at 03:20, Tyler Doyle <notifications@github.com<mailto:notifications@github.com>> wrote:


@irwando<https://github.com/irwando> @RussTheAerialist<https://github.com/RussTheAerialist> the ordering of elements shouldn't matter -- this seems like a Server-side problem?

**Comment by slewt** [comment]
I'm potentially seeing the same thing. When publishing to 2018.2 with server.version set to '3.0', I often get a BadRequest response from the server. If I catch that error, change the server version to '2.4' with the same DatasourceItem, then it goes through. 

Other oddity that I haven't fully tested out: it appears that if I have multiple datasources to publish, all will fail with server.version set to 3.0. After adding the catch above on the first datasource and getting a successful publish, subsequent publish attempts, never get the BadRequest response. Things I know I haven't ruled out are publishing to different sites (using different auth tokens) and enough testing to be sure that it's not random. Pretty sure the first datasource always failed at server version set to '3.0'

**Comment by Bartman0** [comment]
Care to share code snippets?
Much appreciated.



Richard Kooijman
Inergy
Pelmolenlaan 19  • 3447 GW Woerden • T: +31 348 457 666 • M: +31 631 673 387 • www.inergy.nl
Meld je hier aan | Event l Inspiriment op 21 maart
Bekijk hier | Webinar l Dynamic Pricing op 24 januari
Blog | Data analytics in de sport: stilstaan is achterblijven
On 14 Feb 2019, at 12:40, slewt <notifications@github.com<mailto:notifications@github.com>> wrote:


I'm potentially seeing the same thing. When publishing to 2018.2 with server.version set to '3.0', I often get a BadRequest response from the server. If I catch that error, change the server version to '2.4' with the same DatasourceItem, then it goes through.

Other oddity that I haven't fully tested out: it appears that if I have multiple datasources to publish, all will fail with server.version set to 3.0. After adding the catch above on the first datasource and getting a successful publish, subsequent publish attempts, never get the BadRequest response. Things I know I haven't ruled out are publishing to different sites (using different auth tokens) and enough testing to be sure that it's not random. Pretty sure the first datasource always failed at server version set to '3.0'

**Comment by slewt** [comment]
The relevant snippets: 
```python

server = tsc.Server(server_address=XXXX)
server.version = '3.0'

with server.auth.sign_in(auth_req=tableau_creds): 
    # code to get project location and create DatasourceItem
    try:
        new_datasource = server.datasources.publish(datasource_item=datasource, file_path=filename, mode='Overwrite')
    except tsc.server.endpoint.exceptions.ServerResponseError:
        server.version = '2.4'
        new_datasource = server.datasources.publish(datasource_item=datasource, file_path=filename, mode='Overwrite')    

# remainder of method
```

FYI: went with 2.4 as that is what we had been using successfully in the past. I have not tried intermediate api versions.

**Comment by jacalata** [comment]
Order does matter in xml for elements in a sequence. I can repro this locally with newer versions of Server as well. 
https://stackoverflow.com/questions/4328867/does-xml-care-about-the-order-of-elements

**Comment by t8y8** [comment]
Our responses are just elements as constructed by various serializers in Server, but also whatever language you happen to use (python here) -- I don't recommend that we make order part of the API contract, it could change if the underlying java library changes how it writes data out. If you use the JSON representation your even more likely to hit different element ordering because many hash maps/dicts aren't ordered.

**Comment by bcantoni** [comment]
I don't believe there is anything to fix here, but will leave open to investigate that question about changing things in the XSD.

**Comment by jacalata** [comment]
The xsd sequence element generates this restriction on the Java class: 
@XmlType(name = "customViewType", propOrder = {
    "view",
    "workbook",
    "owner"
})

however this is only expected to affect the serialized output, not the acceptable input. 

AFAICT, we could replace xs:sequence with xs:all in every case and the only effect should be removing this (not-supposed-to-exist) constraint on order of elements.

**Comment by jacalata** [comment]
Closing as not-planned. Reviewed in 2026: TSC's `request_factory.py::DatasourceRequest._generate_xml` still emits `<project>` before `<connectionCredentials>`, which doesn't match the XSD sequence the original report validated against — that hasn't changed. What has changed is that Tableau Server no longer rejects the request over element order; `samples/publish_datasource.py` (which exercises this exact XML with `--conn-username` / `--conn-password`) has been shipping and working in the field for years, and there have been no re-reports of a 400011 from ordering since 2019. If a strict-XSD client on the response side is running into a related mismatch, please open a fresh issue with the specific element sequence and a repro against current Server.
